File: 백준/Gold/16236. 아기 상어/아기 상어.py
from sys import stdin
from collections import deque

n = int(stdin.readline())
spaces = [[100] * (n+2)] + [[100] + list(map(int, stdin.readline().split())) + [100] for _ in range(n)] + [[100] * (n+2)]
MOVES = [(-1, 0), (0, -1), (0, 1), (1, 0)]

for y, row in enumerate(spaces):
    if 9 in row:
        shark_pos = (y, row.index(9), 0)
        spaces[shark_pos[0]][shark_pos[1]] = 0
        break

shark_level, eating_count = 2, 0
deq = deque()
visited = [[False] * (n+2) for _ in range(n+2)]

while True:
    visited[shark_pos[0]][shark_pos[1]] = True
    deq.append(shark_pos)
    eating_list = []
    while deq:
        y, x, second = deq.popleft()
        second += 1
        for dy, dx in MOVES:
            dy, dx = dy + y, dx + x
            if not visited[dy][dx]:
                if 0 < spaces[dy][dx] < shark_level: # 잡아먹을 수 있는게 있다면
                    eating_list.append((dy, dx, second))
                    visited[dy][dx] = True
                elif spaces[dy][dx] in (shark_level, 0): # 지나갈 수 있다면
                    visited[dy][dx] = True
                    deq.append((dy, dx, second))
    if not eating_list:
        break
    eating_list.sort(key=lambda pos: (pos[2], pos[0], pos[1]))
    cur_y, cur_x, second = eating_list[0]
    shark_pos = (cur_y, cur_x, second)
    eating_count += 1
    spaces[cur_y][cur_x] = 0
    deq.clear()
    if eating_count == shark_level:
        shark_level += 1
        eating_count = 0
    visited = [[False] * (n+2) for _ in range(n+2)]
    eating_list.clear()
    

            
print(shark_pos[2])

File: 백준/Gold/16236. 아기 상어/아기 상어.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.valueOf(br.readLine().strip());
        int[][] spaces = new int[n+2][];
        spaces[0] = new int[n+2];
        spaces[n+1] = new int[n+2];
        Arrays.fill(spaces[0], 100);
        Arrays.fill(spaces[n+1], 100);
        for (int i = 1; i <= n; i++) {
            spaces[i] = Arrays.stream(("100 " + br.readLine() + " 100").split(" ")).mapToInt(Integer::valueOf).toArray();
        }
        int[][] MOVES = {{-1, 0}, {0, -1}, {0, 1}, {1, 0}};

        int[] sharkPos = new int[3];
        boolean find = false;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++)
                if (spaces[i][j] == 9) {
                    sharkPos = new int[]{i, j, 0};
                    spaces[i][j] = 0;
                    find = true;
                    break;
                }
            if (find) {
                break;
            }
        }
        int sharkLevel = 2, eatingCount = 0;
        ArrayDeque<int[]> deq = new ArrayDeque<>();
        boolean[][] visited = new boolean[n+2][n+2];

        while(true) {
            visited[sharkPos[0]][sharkPos[1]] = true;
            deq.add(sharkPos);
            ArrayList<int[]> eatingList = new ArrayList<>();
            while (!deq.isEmpty()) {
                int[] infos = deq.poll();
                infos[2] += 1;
                for (int[] mv:MOVES) {
                    int dy = mv[0] + infos[0];
                    int dx = mv[1] + infos[1];
                    if (!visited[dy][dx]) {
                        if (0 < spaces[dy][dx] && spaces[dy][dx] < sharkLevel) {
                            eatingList.add(new int[] {dy, dx, infos[2]});
                            visited[dy][dx] = true;
                        } else if (spaces[dy][dx] == sharkLevel || spaces[dy][dx] == 0) {
                            visited[dy][dx] = true;
                            deq.add(new int[]{dy, dx, infos[2]});
                        }
                    }
                }
            }
            if (eatingList.isEmpty()) {
                break;
            }
            eatingList.sort((a, b) -> {
                if (a[2] != b[2]) {
                    return a[2] - b[2];
                } else if (a[0] != b[0]) {
                    return a[0] - b[0];
                } else {
                    return a[1] - b[1];
                }
            });
            int[] curs = eatingList.get(0);
            sharkPos = curs;
            eatingCount += 1;
            spaces[curs[0]][curs[1]] = 0;
            deq.clear();
            if (eatingCount == sharkLevel) {
                sharkLevel += 1;
                eatingCount = 0;
            }
            visited = new boolean[n+2][n+2];
        }

        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(String.valueOf(sharkPos[2] - 1));
        bw.close();
    }
}